servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherBuffer.java

```diff
@@ -249,57 +290,111 @@ public void onComplete() {
+                                final Accumulator<T, B> accumulator = it.accumulator.size() > 0 &&
+                                        pendingUpdater.decrementAndGet(this) >= 0L ? it.accumulator : null;
+                                terminateTarget(accumulator, target, it.terminalNotification, bSubscription);
+                            }
+                        }
                         return;
                     }
                 }
             }
         }
 
-        <T, B> void itemsTerminated(final TerminalNotification terminalNotification) {
+        <T, B> void itemsTerminated(final TerminalNotification terminalNotification,
+                                    final Subscriber<? super B> target, final Cancellable bCancellable) {
             for (;;) {
                 final Object cMaybeAccumulator = maybeAccumulator;
+                // 'itemsTerminated' and 'accumulate' are expected to be sequential:
+                assert cMaybeAccumulator != ADDING;
                 if (cMaybeAccumulator == TERMINATED) {
                     return;
-                }
-                assert !(cMaybeAccumulator instanceof ItemsTerminated);
-
-                ItemsTerminated<T, B> itemsTerminated = new ItemsTerminated<>(cMaybeAccumulator, terminalNotification);
-                if (maybeAccumulatorUpdater.compareAndSet(this, cMaybeAccumulator, itemsTerminated)) {
+                } else if (cMaybeAccumulator != null &&
+                        NextAccumulatorHolder.class.equals(cMaybeAccumulator.getClass())) {
+                    // Terminated while `nextAccumulator` delivers onNext, notify that thread using ItemsTerminated
+                    // wrapper. There is no more concurrency because neither `itemsTerminated` nor `accumulate` will be
+                    // invoked again.
+                    @SuppressWarnings("unchecked")
+                    final ItemsTerminated<T, B> itemsTerminated = new ItemsTerminated<>(
+                            (NextAccumulatorHolder<T, B>) cMaybeAccumulator, terminalNotification);
+                    if (maybeAccumulatorUpdater.compareAndSet(this, cMaybeAccumulator, itemsTerminated)) {
+                        return;
+                    }
+                } else if (maybeAccumulatorUpdater.compareAndSet(this, cMaybeAccumulator, TERMINATED)) {
+                    assert cMaybeAccumulator instanceof CountingAccumulator;
+                    // Deliver the last boundary only if there are some items pending and demand is present
+                    @SuppressWarnings("unchecked")
+                    final CountingAccumulator<T, B> accumulator = (CountingAccumulator<T, B>) cMaybeAccumulator;
+                    terminateTarget(accumulator.size() > 0 && pendingUpdater.decrementAndGet(this) >= 0 ?
+                            accumulator : null, target, terminalNotification, bCancellable);
```

Comment:
I was experimenting with the work and this test doesn't behave as expected

```
    @Test
    void test() throws InterruptedException {
        final AtomicInteger idSeq = new AtomicInteger();
        final CountDownLatch completion = new CountDownLatch(1);

        Publisher<Integer> tPublisher = Publisher.range(1, 1000);
        Publisher<Accumulator<Integer, Integer>> boundaries = new Publisher<Accumulator<Integer, Integer>>() {
            @Override
            protected void handleSubscribe(final Subscriber<? super Accumulator<Integer, Integer>> subscriber) {
                subscriber.onSubscribe(new Subscription() {
                    @Override
                    public void request(final long n) {
                        int id = idSeq.incrementAndGet();
                        subscriber.onNext(new Accumulator<Integer, Integer>() {

                            private int sum;

                            @Override
                            public void accumulate(@Nullable final Integer item) {
                                System.err.println("Accumulating on " + id + " " + item + " sum: " + sum);
                                sum += item;
                            }

                            @Override
                            public Integer finish() {
                                System.err.println("Sum for " + id + " is: " + sum);
                                return sum;
                            }
                        });
                    }

                    @Override
                    public void cancel() {
                        //ignore
                    }
                });
            }
        };

        BlockingQueue<Integer> buffers = new LinkedBlockingDeque<>();
        toSource(tPublisher.buffer(new TestBufferStrategy(boundaries, 3)))
                .subscribe(new Subscriber<Integer>() {
                    @Override
                    public void onSubscribe(Subscription s) {
                        s.request(1);
                    }

                    @Override
                    public void onNext(@Nullable Integer integer) {
                        assert integer != null;
                        System.err.println("Downstream received " + integer);
                        buffers.add(integer);
                    }

                    @Override
                    public void onError(Throwable t) {
                        t.printStackTrace();
                    }

                    @Override
                    public void onComplete() {
                        completion.countDown();
                    }
                });

        completion.await();
    }
```

If you only return 1 boundary it works, if you allow it to return 2 (as is now) then the 1st boundary will return a 0 sum, and the 2nd one doesn't emit (even though it *does* accumulate).

Reply:
Discussed offline:
1. The last empty accumulation is discarded because a user usually does not request it. Internally we request at least +1 boundary from what the user requested. It helps to avoid confusion when users accumulate 3 items of `Publisher.from(1, 2, 3)`. Instead of receiving 2 accumulations, they will get one.
2. This is a bit different compare to empty accumulations in between because those are usually driven by a timer. With count-based accumulation users won't see empty accumulations.
3. Clarified in javadoc under what conditions boundaries can be discarded and that the `finish` method won't be invoked.